persist/encoding/stream.go

```diff
@@ -0,0 +1,148 @@
+// Copyright (c) 2017 Uber Technologies, Inc
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE
+
+package encoding
+
+import (
+	"bytes"
+	"fmt"
+	"io"
+)
+
+// DecoderStream is a data stream that is read by the decoder,
+// it takes both a reader and the underlying backing bytes.
+// This is constructed this way since the decoder needs access
+// to the backing bytes when taking refs directly for decoding byte
+// slices without allocating bytes itself but also needs to progress
+// the reader (for instance when a reader is a ReaderWithDigest that
+// is calculating a digest as its being read).
+type DecoderStream interface {
+	io.Reader
+
+	// Reset resets the decoder stream for decoding a new byte slice.
+	Reset(b []byte)
+
+	// ReadByte reads the next byte.
+	ReadByte() (byte, error)
+
+	// UnreadByte unreads the last read byte or returns error if none read
+	// yet. Only a single byte can be unread at a time, a consecutive call
+	// to UnreadByte will result in an error.
+	UnreadByte() error
```

Comment:
One thing that's weird about this is the implementation of calling `UnreadByte()` captures the unread byte in the decoderStream. So calls to `Bytes()` right after a successful `UnreadByte()` return a byte slice without the "unread byte". Is that kosher?

Reply:
`Bytes()` actually always gives you the entire underlying byte slice without any offsets.  `Remaining()` which is used to calculate position into the current byte slice however does not take into account whether you have unread a byte or not, so that is a good observation.  I'll fix this.